x-pack/plugin/rollup/src/test/java/org/elasticsearch/xpack/downsample/DataStreamTests.java

```diff
@@ -0,0 +1,225 @@
+import org.elasticsearch.cluster.metadata.DataStreamAction;
+import org.elasticsearch.cluster.metadata.IndexMetadata;
+import org.elasticsearch.cluster.metadata.Template;
+import org.elasticsearch.common.compress.CompressedXContent;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.datastreams.DataStreamsPlugin;
+import org.elasticsearch.index.Index;
+import org.elasticsearch.index.IndexMode;
+import org.elasticsearch.index.IndexSettings;
+import org.elasticsearch.index.mapper.DateFieldMapper;
+import org.elasticsearch.index.query.MatchAllQueryBuilder;
+import org.elasticsearch.plugins.Plugin;
+import org.elasticsearch.rest.RestStatus;
+import org.elasticsearch.search.aggregations.bucket.histogram.DateHistogramAggregationBuilder;
+import org.elasticsearch.search.aggregations.bucket.histogram.DateHistogramInterval;
+import org.elasticsearch.search.aggregations.bucket.histogram.InternalDateHistogram;
+import org.elasticsearch.search.builder.SearchSourceBuilder;
+import org.elasticsearch.test.ESIntegTestCase;
+import org.elasticsearch.xcontent.XContentType;
+import org.elasticsearch.xpack.core.downsample.DownsampleAction;
+import org.elasticsearch.xpack.core.downsample.DownsampleConfig;
+import org.elasticsearch.xpack.rollup.Rollup;
+import org.hamcrest.Matchers;
+
+import java.io.IOException;
+import java.util.Collection;
+import java.util.List;
+import java.util.Locale;
+import java.util.concurrent.ExecutionException;
+
+import static org.elasticsearch.cluster.metadata.MetadataIndexTemplateService.DEFAULT_TIMESTAMP_FIELD;
+import static org.hamcrest.Matchers.equalTo;
+
+public class DataStreamTests extends ESIntegTestCase {
+
+    private static final long A_BIT_MORE_THAN_3_HOURS_MILLIS = 10_000_000L;
+
+    @Override
+    protected Collection<Class<? extends Plugin>> nodePlugins() {
+        return List.of(Rollup.class, DataStreamsPlugin.class);
+    }
+
+    public void testDataStreamDownsample() throws ExecutionException, InterruptedException, IOException {
+        // GIVEN
+        final String dataStreamName = randomAlphaOfLength(5).toLowerCase(Locale.ROOT);
+        putComposableIndexTemplate("1", List.of(dataStreamName));
+        final CreateDataStreamAction.Request request = new CreateDataStreamAction.Request(dataStreamName);
+        client().execute(CreateDataStreamAction.INSTANCE, request).actionGet();
+        indexDocs(dataStreamName, 10, 0L);
+        final RolloverResponse rolloverResponse = client().admin().indices().rolloverIndex(new RolloverRequest(dataStreamName, null)).get();
+        // NOTE: here delay = 10_000_000 because the next data stream write index is created with a start time of (about) two hours in the
+        // future.
+        // 10_000_000 milliseconds is a bit more than 3 hours, which is more than enough to make sure we are not writing documents to the
+        // old data stream index.
+        indexDocs(dataStreamName, 10, A_BIT_MORE_THAN_3_HOURS_MILLIS);
```

Comment:
maybe there is a way to programmatically get the `start_time` of the new backing index? Will check that.

Reply:
Yes, maybe use the get index settings api to get the start_time setting? And then just increment 1 ms for each document to be indexed?

Reply:
So instead just pass the start time to `indexDocs(...)` method? And then iteratively add a ms to the start time and use that as timestamp for the document to be indexed?